File: src/main/java/com/github/cao/awa/conium/components/ConiumComponentType.java
package com.github.cao.awa.conium.components;

import com.github.cao.awa.conium.components.value.ConiumValueCreator;
import com.mojang.serialization.Codec;
import net.minecraft.component.ComponentType;
import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.registry.Registries;
import net.minecraft.util.Util;
import org.jetbrains.annotations.Nullable;

public record ConiumComponentType<T>(Codec<T> codec,
                                     PacketCodec<? super RegistryByteBuf, T> packetCodec,
                                     ConiumValueCreator<T> valueCreator,
                                     String type
) implements ComponentType<T> {
    public String toString() {
        return Util.registryValueToString(Registries.DATA_COMPONENT_TYPE, this);
    }

    @Override
    public @Nullable Codec<T> getCodec() {
        return this.codec;
    }

    @Override
    public PacketCodec<? super RegistryByteBuf, T> getPacketCodec() {
        return this.packetCodec;
    }
}

File: common/src/main/kotlin/com/github/cao/awa/conium/component/ConiumComponentType.kt
package com.github.cao.awa.conium.component

import com.github.cao.awa.conium.component.value.ConiumValueCreator
import com.mojang.serialization.Codec
import net.minecraft.component.ComponentType
import net.minecraft.network.RegistryByteBuf
import net.minecraft.network.codec.PacketCodec
import net.minecraft.registry.Registries
import net.minecraft.util.Util

@JvmRecord
data class ConiumComponentType<T>(
    val xCodec: Codec<T>?,
    val packetCodec: PacketCodec<in RegistryByteBuf, T>?,
    val valueCreator: ConiumValueCreator<T>,
    val type: String
) : ComponentType<T> {
    override fun toString(): String = Util.registryValueToString(Registries.DATA_COMPONENT_TYPE, this)

    override fun getCodec(): Codec<T>? = this.xCodec

    override fun getPacketCodec(): PacketCodec<in RegistryByteBuf, T>? = this.packetCodec
}